frappe/integrations/doctype/google_contacts/google_contacts.py

```diff
@@ -0,0 +1,110 @@
+REQUEST = 'https://people.googleapis.com/v1/people/me/connections'
+
+class GoogleContacts(Document):
+
+	def get_access_token(self):
+		if not self.refresh_token:
+			raise frappe.ValidationError(_("Google Contacts is not configured."))
+
+		data = {
+			'client_id': self.client_id,
+			'client_secret': self.get_password(fieldname='client_secret', raise_exception=False),
+			'refresh_token': self.get_password(fieldname='refresh_token', raise_exception=False),
+			'grant_type': "refresh_token",
+			'scope': SCOPES
+		}
+
+		try:
+			r = requests.post('https://www.googleapis.com/oauth2/v4/token', data=data).json()
+		except requests.exceptions.HTTPError:
+			frappe.throw(_("Something went wrong during the token generation. Please request again an authorization code."))
+
+		return r.get('access_token')
+
+@frappe.whitelist()
+def google_callback(code=None):
+	doc = frappe.get_doc("Google Contacts")
+
+	redirect_uri = get_request_site_address(True) + "?cmd=frappe.integrations.doctype.google_contacts.google_contacts.google_callback"
+
+	if code is None:
+		return {
+			'url': 'https://accounts.google.com/o/oauth2/v2/auth?access_type=offline&response_type=code&prompt=consent&client_id={}&include_granted_scopes=true&scope={}&redirect_uri={}'.format(doc.client_id, SCOPES, redirect_uri)
+		}
+	else:
+		try:
+			data = {
+				'code': code,
+				'client_id': doc.client_id,
+				'client_secret': doc.get_password(fieldname='client_secret', raise_exception=False),
+				'redirect_uri': redirect_uri,
+				'grant_type': 'authorization_code'
+			}
+			r = requests.post('https://www.googleapis.com/oauth2/v4/token', data=data).json()
+			frappe.db.set_value("Google Contacts", None, "authorization_code", code)
+
+			if 'refresh_token' in r:
+				frappe.db.set_value("Google Contacts", None, "refresh_token", r.get("refresh_token"))
+
+			frappe.db.commit()
+			frappe.local.response["type"] = "redirect"
+			frappe.local.response["location"] = "/desk#Form/Google%20Contacts"
+			return
+		except Exception as e:
+			frappe.throw(e.message)
+
+@frappe.whitelist()
+def sync():
+	doc = frappe.get_doc("Google Contacts")
+	access_token = doc.get_access_token()
+
+	params = {'personFields': 'names,emailAddresses'}
+	headers = {'Authorization': 'Bearer {}'.format(access_token)}
+
+	try:
+		r = requests.get(REQUEST, headers=headers, params=params)
+	except Exception as e:
+		frappe.throw(e.message)
+
+	try:
+		r = r.json()
+	except Exception as e:
+		# if request doesn't return json show HTML ask permissions or to identify the error on google side
+		frappe.throw(e)
+
+	connections = r.get('connections')
+	for idx, connection in enumerate(connections):
+		for name in connection.get('names'):
+			show_progress(len(connections), "Google Contacts", idx, name.get('displayName'))
+			for email in connection.get('emailAddresses'):
+				if not frappe.db.exists("Contact", {"email_id": email.get('value')}):
+					contact = frappe.get_doc({
```

Comment:
add a `source` property in Contact and and set "Google Contacts" so this is clear where this was imported from

Reply:
Done